exporter/otelarrowexporter/metadata.go

```diff
@@ -0,0 +1,201 @@
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"go.opentelemetry.io/collector/client"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/otel/attribute"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	// errTooManyExporters is returned when the MetadataCardinalityLimit has been reached.
+	errTooManyExporters = consumererror.NewPermanent(errors.New("too many exporter metadata-value combinations"))
+	// errUnexpectedType is returned when the object in the map isn't the expected type
+	errUnexpectedType = errors.New("unexpected type in map")
+)
+
+type metadataExporter struct {
+	config   *Config
+	settings exporter.Settings
+	scf      streamClientFactory
+	host     component.Host
+
+	metadataKeys []string
+	exporters    sync.Map
+
+	// Guards the size and the storing logic to ensure no more than limit items are stored.
+	// If we are willing to allow "some" extra items than the limit this can be removed and size can be made atomic.
+	lock sync.Mutex
+	size int
+}
+
+var _ exp = (*metadataExporter)(nil)
+
+func newMetadataExporter(cfg component.Config, set exporter.Settings, streamClientFactory streamClientFactory) (exp, error) {
+	oCfg := cfg.(*Config)
+	// use lower-case, to be consistent with http/2 headers.
+	mks := make([]string, len(oCfg.MetadataKeys))
+	for i, k := range oCfg.MetadataKeys {
+		mks[i] = strings.ToLower(k)
+	}
+	sort.Strings(mks)
+	if len(mks) == 0 {
+		return newExporter(cfg, set, streamClientFactory)
+	}
+	return &metadataExporter{
+		config:       oCfg,
+		settings:     set,
+		scf:          streamClientFactory,
+		metadataKeys: mks,
+	}, nil
+}
+
+func (e *metadataExporter) getSettings() exporter.Settings {
+	return e.settings
+}
+
+func (e *metadataExporter) getConfig() component.Config {
+	return e.config
+}
+
+func (e *metadataExporter) helperOptions() []exporterhelper.Option {
+	return []exporterhelper.Option{
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		exporterhelper.WithTimeout(e.config.TimeoutSettings),
+		exporterhelper.WithRetry(e.config.RetryConfig),
+		exporterhelper.WithQueue(e.config.QueueSettings),
+		exporterhelper.WithStart(e.start),
+		exporterhelper.WithShutdown(e.shutdown),
+	}
+}
+
+func (e *metadataExporter) start(_ context.Context, host component.Host) (err error) {
+	e.host = host
+	return nil
+}
+
+func (e *metadataExporter) shutdown(ctx context.Context) error {
+	var err error
+	e.exporters.Range(func(_ any, value any) bool {
+		be, ok := value.(exp)
+		if !ok {
+			err = multierr.Append(err, fmt.Errorf("%w: %T", errUnexpectedType, value))
+			return true
+		}
```

Comment:
Same, I don't think this is possible--otherwise it's an internal bug that unittesting would uncover.

Reply:
Removed